experiment_handler/experiment_handler.py

```diff
@@ -98,37 +98,41 @@ def retrieve_experiment():
     experiment = args.experiment_module
     experiment_mod = importlib.import_module("experiments." + experiment)
 
-    experiment_classes = {}
-    for class_name, obj in inspect.getmembers(experiment_mod, inspect.isclass):
-        experiment_classes[class_name] = obj
-
-    # need to have one ExperimentPrototype and one user-specified class.
-    try:
-        experiment_proto_class = experiment_classes['ExperimentPrototype']
-        del experiment_classes['ExperimentPrototype']
-    except KeyError:
-        errmsg = "Your experiment is not built from parent class ExperimentPrototype - exiting"
-        raise ExperimentException(errmsg)
-
-    list_experiments = []
-    for class_name, class_obj in experiment_classes.items():
-        if experiment_proto_class in inspect.getmro(class_obj):  # an experiment
-            # must inherit from ExperimentPrototype
+    # find the class or classes *defined* in this module.
+    # returns list of class name and object
+    experiment_classes = [(m[0], m[1]) for m in inspect.getmembers(
+                          experiment_mod, inspect.isclass) if
+                          m[1].__module__ == experiment_mod.__name__]
+
+    # remove any classes that do not have ExperimentPrototype as parent.
+    for (class_name, class_obj) in experiment_classes:
+        if ExperimentPrototype not in inspect.getmro(class_obj):
+            # an experiment must inherit from ExperimentPrototype
             # other utility classes might be in the file but we will ignore them.
-            list_experiments.append(class_obj)
+            experiment_classes.remove((class_name, class_obj))
 
-    if len(list_experiments) != 1:
-        errmsg = "You have zero or more than one experiment class in your experiment " \
-                 "file - exiting"
+    # experiment_classes should now only have classes *defined* in the module,
+    # that have ExperimentPrototype as parent.
+    if len(experiment_classes) == 0:
+        errmsg = "No experiment classes are present that are built from"\
+                 " parent class ExperimentPrototype - exiting"
+        raise ExperimentException(errmsg)
+    if len(experiment_classes) > 1:
+        errmsg = "You have more than one experiment class in your " \
+                 "experiment file - exiting"
         raise ExperimentException(errmsg)
 
-    Experiment = list_experiments[0]  # this is the experiment class that we need to run.
+    # this is the experiment class that we need to run.
+    Experiment = experiment_classes[0][1]
 
+    printing('Retrieving experiment: {} from module {}'.format(
+             experiment_classes[0][0], experiment_mod))
     try:
         return Experiment
     except NameError:
-        errmsg = "Cannot find the experiment inside your module. Please make sure there is a " \
-                 "class that inherits from ExperimentPrototype in your module."
+        errmsg = "Something went wrong: Cannot find the experiment inside " \
```

Comment:
Has this ever happened? Under what circumstances?

Reply:
I don't remember - so I left it in case it does.